File: examples/i2c/aht/c/aht.c


#include "aht.h"
#include <stdio.h>
#include <time.h>

#define ARRAY_SIZE(arr) (sizeof(arr) / sizeof(*arr))

// commands
#define AHT_INIT_CMD    0xE1
#define AHT_SOFT_RESET  0xBA
#define AHT_TRIG_MEAS   0xAC
#define AHT_DAT1_CMD    0x33
#define AHT_DAT2_CMD    0x00

#define aht_log         printf

// Commands to trigger measurement
static const uint8_t sht_cmd_trig[3]  = {AHT_TRIG_MEAS, AHT_DAT1_CMD, AHT_DAT2_CMD};
// Calibration commands
static const uint8_t sht_cmd_calib[3] = {0xE1, 0x08, 0x00};

static int fd = -1;

bool aht_init(const char* device, uint8_t address)
{
    // Open I2C Bus
    if ((fd = open(device, O_RDWR)) < 0)
    {
        aht_log("Failed to open the i2c bus");
        return false;
    }

    if (ioctl(fd, I2C_SLAVE, address) < 0)
    {
        aht_log("Failed to acquire bus access and/or talk to slave.\n");
        return false;
    }

    // Calibrate
    if (write(fd, sht_cmd_calib, ARRAY_SIZE(sht_cmd_calib)) != ARRAY_SIZE(sht_cmd_calib))
    {
        aht_log("Failed to write to the i2c bus.\n");
        return false;
    }

    return true;
}

bool aht_read(float* temp, float* humi)
{
    uint8_t rxbuf[6];

    // Send trigger command
    if (write(fd, sht_cmd_trig, ARRAY_SIZE(sht_cmd_trig)) != ARRAY_SIZE(sht_cmd_trig))
    {
        aht_log("Failed to write to the i2c bus.\n");
        return false;
    }

    // The sensor needs 75 ms before reading the data
    aht_delay(75);

    // Read data
    if (read(fd, rxbuf, ARRAY_SIZE(rxbuf)) != ARRAY_SIZE(rxbuf))
    {
        aht_log("Failed to read from the i2c bus.\n");
        return false;
    }

    // check if the device is busy and try to read data again
    while (rxbuf[0] & (1 << 7))
    {
        if (read(fd, rxbuf, ARRAY_SIZE(rxbuf)) != ARRAY_SIZE(rxbuf))
        {
            aht_log("Failed to read from the i2c bus.\n");
            return false;
        }
    }

    uint32_t humi_raw = ((uint32_t)rxbuf[1] << 12) | ((uint16_t)rxbuf[2] << 4) | (rxbuf[3] >> 4);
    uint32_t temp_raw = (((uint32_t)rxbuf[3] & 0x0F) << 16) | ((uint16_t)rxbuf[4] << 8) | (rxbuf[5]);

    *humi = (((float)humi_raw) / (1 << 20)) * 100;
    *temp = ((((float)temp_raw) / (1 << 20)) * 200) - 50;

    return true;
}

void aht_delay(long ms)
{
    long    pause;
    clock_t now, then;
    pause = ms * (CLOCKS_PER_SEC / 1000);
    now = then = clock();
    while ((now - then) < pause)
    {
        now = clock();
    }
}

File: examples/i2c/aht/python/aht.py
import smbus
import time

class AHTxx:
    def __init__(self, bus=1, address=0x38):
        self.bus = smbus.SMBus(bus)
        self.address = address
        
        time.sleep(0.04)

        # soft reset the sensor
        self.bus.write_byte(self.address, 0x0ba)
        time.sleep(0.02)

        # check if the sensor is calibrated
        if not self._get_status() & 0x08:
            self.calibrate()

    def _get_status(self):
        self.bus.write_byte(self.address, 0x71)
        return self.bus.read_byte(self.address)

    def _calibrate(self):
        self.bus.write_i2c_block_data(self.address, 0xbe, [0x08, 0x00])
        time.sleep(0.01)
        self._wait_while_busy()

    def _wait_while_busy(self):
        while self._get_status() & 0x80:
            time.sleep(0.01)

    def measure(self):
        self.bus.write_i2c_block_data(self.address, 0xac, [0x33, 0x00])
        time.sleep(0.08)
        self._wait_while_busy()

        data = self.bus.read_i2c_block_data(self.address, 0, 7)

        crc = self._crc8(data[:6])

        if crc == data[6]:
            self.humidity = (data[1] << 12)
            self.humidity |= (data[2] << 4)
            self.humidity |= (data[3] >> 4)
            self.humidity /= (1 << 20)
            self.humidity *= 100

            self.temperature = ((data[3] & 0x0f) << 16)
            self.temperature |= (data[4] << 8)
            self.temperature |= data[5]
            self.temperature /= (1 << 20)
            self.temperature *= 200
            self.temperature -= 50
        else:
            self.humidity = None
            self.temperature = None

    def _crc8(self, data):
        crc = 0xff

        for x in data:
            crc ^= x
            for i in range(8):
                if crc & 0x80:
                    crc <<= 1
                    crc ^= 0x31
                else:
                    crc <<= 1
                
        return crc & 0xff

if __name__ == "__main__":
    aht20 = AHTxx(2)
    aht20.measure()

    if aht20.humidity != None:
        print("temperature : %.1f °C" % aht20.temperature)
        print("humidity : %.1f %%" % aht20.humidity)

